A48-xds-least-request-lb-policy.md

```diff
@@ -0,0 +1,225 @@
+algorithm as Envoy's implementation.  However, this policy will be
+implemented in a non-xDS-specific way, so that it can also be used without
+xDS in the future.
+
+##### LB Policy Config
+
+The `least_request_experimental` policy will have the following config:
+
+```proto
+message LeastRequestLoadBalancingConfig {
+  uint32 choice_count = 1;  // Optional, defaults to 2.
+}
+```
+
+The `choice_count` determines the number of randomly sampled subchannels which should be compared
+for each pick by the picker.
+Envoy currently supports any `choice_count` value greater than two.
+In the `least_request_experimental` policy however, only an effective `choice_count` in the
+range `[2, 10]` will be supported. If a `LeastRequestLoadBalancingConfig` with a `choice_count > 10`
+is received, the `least_request_experimental` policy will set `choice_count = 10`.
+
+##### Subchannel state handling
+
+The subchannel state handling implemented in the `round_robin` policy will be re-used for `least_request_experimental`.
+Implementations may break out this subchannel state handling and have `round_robin` and `least_request_experimental`
+inherit the logic instead. This would be similar to the "base balancer" already implemented in grpc-go
+[here](https://github.com/grpc/grpc-go/blob/03268c8ed29e801944a2265a82f240f7c0e1b1c3/balancer/base/balancer.go).
+The subchannel state handling used for the `round_robin` policy can be found in each language-specific implementation:
+
+* [Java](https://github.com/grpc/grpc-java/blob/c1e19af86dea0b9a8725969a95e116029397ad4d/core/src/main/java/io/grpc/util/RoundRobinLoadBalancer.java)
+* [C++](https://github.com/grpc/grpc/blob/79d684529d9db60aa2c5e82deb43e297a6818cfb/src/core/ext/filters/client_channel/lb_policy/round_robin/round_robin.cc)
+* [Go](https://github.com/grpc/grpc-go/blob/03268c8ed29e801944a2265a82f240f7c0e1b1c3/balancer/base/balancer.go)
+
+##### Aggregated Connectivity State
+
+The `least_request_experimental` policy will use the same heuristic for determining the aggregated
+connectivity state as the `round_robin` policy.
+The general rules in order of precedence are:
+
+1. If there is at least one subchannel with the state `READY`, the aggregated connectivity state is `READY`.
+2. If there is at least one subchannel with the state `CONNECTING` or `IDLE`,
+   the aggregated connectivity state is `CONNECTING`.
+3. If all subchannels have the state `TRANSIENT_FAILURE`, the aggregated connectivity state is `TRANSIENT_FAILURE`.
+
+##### Picker Behavior
+
+The picker should be given a list of all subchannels with the `READY` state.
+With this list, the picker should pick a subchannel based on the following pseudo-code:
+
+```
+candidate = null;
+for (int i = 0; i < choiceCount; ++i) {
+    sampled = subchannels[random_integer(0, subchannels.length)];
+    if (candidate == null) {
+        candidate = sampled;
+        continue;
+    }
+
+    if (sampled.active_requests < candidate.active_requests) {
```

Comment:
Where is the policy going to store the count of active requests for each subchannel?

In Java and Go, this will probably be fairly straightforward; the policy can implement its own subchannel wrapper to store this field.  But C-core has a slightly more complicated model for handling subchannels, which will make this a bit more challenging.

In Java and Go, the LB policy is effectively the single owner of the subchannel.  Whenever the LB policy is given an updated list of addresses, it compares that list to the previous list and keeps the subchannels it already had for any addresses that are still present on the new list.  This avoids unnecessary connection churn caused by deleting the existing subchannel and creating a new subchannel for the same address.  And if those subchannels are wrapped by the LB policy, that wrapper will be a stable object for each address that it can use to store the active request count.

In C-core, however, the subchannels are ref-counted and can have multiple owners, and the `SubchannelInterface` object returned to the LB policy is just a thin wrapper that holds a ref to the real subchannel.  The channel maintains a pool of subchannels, and when an LB policy asks the channel to "create" a subchannel for a given address, if the pool already contains a subchannel for that address, then the channel returns a `SubchannelInterface` object that holds a new ref to that existing subchannel instead of creating a new subchannel.  So when an LB policy is given an updated list of addresses, it doesn't bother to compare with the old list; it just unconditionally "creates" a new subchannel for every address in the list and unrefs the subchannels in the old list.  If there are addresses in common, the new ref will have been created before the old ref is released, so there will not wind up being any connection churn.

What this means is that every time the `least_request` policy gets an updated address list, it will wind up replacing the `SubchannelInterface` object for a given address, even though the underlying subchannel is not changing, which means that the subchannel wrapper is not going to be a stable place to keep the active request count.  So we probably need to come up with something different there.

Reply:
I'm not super familiar with the C-core implementation and would need to spend some more time to give a more proper answer to this.
I would (probably naively) think that doing a similar approach to what `XdsClusterImplLb` does for circuit breaking would work reasonably well here as well with some small adjustments.
So the least request implementation would have a `SubchannelCallCounterMap` similar to [CircuitBreakerCallCounterMap](https://github.com/grpc/grpc/blob/79d684529d9db60aa2c5e82deb43e297a6818cfb/src/core/ext/filters/client_channel/lb_policy/xds/xds_cluster_impl.cc#L47-L107) that somehow uses the real subchannel as key to find the corresponding counter.
One downside with this however, is that the lb policy might need to do a diff between re-resolves to clean up counters for removed addresses.
As for churn, it might not be too bad since I guess an address that gets re-introduced wouldn't/shouldn't have any outstanding requests anyway (even though it already might have an open connection since before).

Reply:
Yeah, that's definitely one possibility.  The entries in the map could be handled via simple ref-counts held by wrapping the `SubchannelInterface` objects (and probably also by the per-call completion callbacks that decrement the counters), so they'd clean themselves up automatically when the address stops being used.

But this brings up another question: Do we want these counters to be global?  In the case of circuit breaking, the counters are global, because they're per-cluster limits, and no matter how many gRPC channels are talking to the same xDS cluster, we want to apply the same limits across all of them.  But outside of the xDS space, LB policies are generally per-channel with no state shared across channels, and it would be somewhat surprising if they shared state globally.  It seems like the goal here is to build the `least_request` policy in a way that works for both the xDS and non-xDS use-cases, so we should carefully think through what behavior we want here.

(Note that the `xds_cluster_impl` policy *is* xDS-specific, so this isn't an issue there.  And the `ring_hash` policy, which is *not* xDS-specific, doesn't have this issue, since the same hash algorithm will be used in every channel even without any global state.  So I think this is the first time we're really having to figure out how to bridge this particular gap.)

In the xDS case, we probably would want to have global counters here, because if a given cluster uses the `least_request` policy, we ideally want to consider all traffic going to the cluster (i.e., we would use the same argument that we used to choose global state for circuit breaking).  If we don't do that, then this will be a case where our behavior will differ from Envoy's, so at minimum, we should document that.

In the non-xDS case, it would probably be a bit surprising if two channels started counting each others' traffic.  But on the other hand, it probably wouldn't actually hurt anything, and it might actually wind up yielding better balancing overall.  However, it is a notable difference from how existing policies work.

It's also worth noting that in Java and Go, each channel will have its own set of subchannels, but C-core by default shares subchannels between channels with the same target and channel args.  If we use global counters, that seems a little more intuitive in the C-core model, whereas it seems a little strange in the Java and Go model; if we don't use global counters, then the opposite is true.  I don't think this will cause any actual problem either way; it's just a little counter-intuitive to share a connection but not see all of the call traffic on the connection (or vice-versa).

In terms of implementation, note that if we do want global counters, then we'd probably need a global map similar to the one we use for circuit breaking, even in Java and Go.  We'll probably need to give some thought to exactly what keys we use for the global map; we probably want it to be more than just the address, because we probably don't want to share counters in the case where there are two channels talking to two different services via the same proxy.

@ejona86 or @dfawley, any thoughts on this?

Reply:
As you point out, there are several cases, especially in java/go, where pure global counters probably aren't a good fit.
For example, talking to different services through a proxy or using different subchannels/connections that target the same address.
One other aspect is that xDS-specific information like cluster name isn't even available in this LB layer, so using something like that for the key seems unfeasible.
Instead of defining it as global or non-global, one option could maybe be to just define the counters as being associated to each unique subchannel:

```
Similarly to how underlying subchannels each have a connection state attribute (READY,IDLE,etc.),
each underlying subchannel will be associated with a single counter for outstanding requests.
This may be implemented either using custom subchannel attributes or a global map that maps the subchannel
memory address to it's corresponding counter.
```

This means that in C-core, counters would be shared equivalently to how subchannels themselves are shared, while in Java/Go, the counters would never be shared since subchannels are never shared.

While this approach wouldn't optimize well for certain edge-cases, it's a fairly simple approach to reason about or debug if there ever are any LB issues.
With the circuit breaking and fully global counters, it's much easier in this regard since hitting the circuit breaker limits gives feedback in the form of an RPC error whereas LBs will be completely silent in all cases.

Reply:
I'd like feedback from Eric and Doug before making a call here.  Personally, I'm leaning in the direction of global counters, since there are some advantages and there isn't really any compelling disadvantage; the behavior may be a little surprising, but it seems like it wouldn't actually hurt anything in common cases.

> One other aspect is that xDS-specific information like cluster name isn't even available in this LB layer, so using something like that for the key seems unfeasible.

One possible way to address that might be by having the `least_request` policy define a per-address attribute indicating some additional value to include in the key for the global map.  We could then have the xDS code pass the cluster name in using that attribute.

Reply:
TL;DR: My suggestion is the last (non-parenthetical) paragraph. Unfortunately it took this mini-proof to arrive there; no summary for you!

I think subchannel sharing is a red herring. C core can do the same behavior of Go and Java by simply doing the address-level diffing that Go and Java do; it can "manually" reassociate the counters from the old subchannel to the new subchannel when it gets address updates.

I think that's a separate discussion from "what scope should the counters be." My initial gut reaction was also they should be associated with the subchannel, and if we choose to do anything we might break subchannel sharing (e.g., by including the cluster name in channel args). But I think that is neither here nor there. With per-subchannel counts 1) we'd still lose any RPC counts when the policy using the subchannel isn't least-request, and 2) there's nothing all that special about _the connection_ as if the connection receives GOAWAY we'd sum its counts with any RPCs on a new connection.

This policy is about spreading load and reducing load to slower backends or backends that may have unluckily received more costly requests. There's no need for perfect global information here, as this client doesn't know the number of requests sent by other clients. There are probably multiple valid scopes for the counters from a correctness perspective, and so it is just an optimization problem when choosing between them.

Note that the policy semi-assumes a reasonable number of concurrent RPCs, otherwise it devolves into random. It doesn't _require_ many concurrent RPCs, but there's no benefit to using it over random/round_robin for low concurrent RPCs.

If we take a least-request policy with a fixed set of addresses and duplicate the entire thing N times (duplicating the counters as well), then the behavior of those N policies should actually be fairly similar to the single unified policy. That only devolves if the number of concurrent RPCs is low. If each policy receives a different type of traffic (different methods, with different cost per method), then that still holds; there can be small interactions between the different types of traffic, but they should be in the noise with sufficient concurrent RPCs. (For lower concurrent RPCs, separate policies might be _better_, as that will distribute each _type_ of RPC more evenly.)

The behavior would change if the N policies have different addresses. Assuming our target is to emulate Envoy's per-cluster-per-address scope, this means that per-shared-subchannel counters is likely a bad idea, but per-shared-subchannel-per-cluster would be okay (if we assume XdsClient sharing and use of EDS, to guarantee the same addresses in each policy).

I don't see a need for global state in Java and Go. In C, I think we should just select the easiest-to-implement approach, whether that is to "manually" transfer counts during address updates or to use the cluster and subchannel as keys into a global map. Assuming similar addresses provided to each LB instance, both approaches should have similar enough behavior that I think the main dimension to optimize is effort. If we don't think that is a safe assumption (e.g., not using EDS), then it is probably best for C to do the "manual" diffing to transfer counts similar to Go and Java.

(If we do go with a cluster+subchannel global map in C, I'd hope we design it in an xds-agnostic way so the policy can still be used outside of xds at some point. I think there are ways to do that easily, so I'm not discussing it here.)

Reply:
I somewhat agree with Eric, though I would say that aggregating as many sources as possible will yield the best results.  I.e. if you had a global view, you could direct traffic most efficiently.  If you have a full view of the current client _machine_, you could direct traffic a bit less efficiently.  If you limit it to the current client _process_, you lose some efficiency.  And if you limit it to a single client _instance_, you lose even more efficiency.  This goes with what Eric said that "the policy semi-assumes a reasonable number of concurrent RPCs" -- with more sources of RPCs, we will have more concurrent RPCs.

So, on average, the policy should give the best results if implemented as globally as possible.  If it was going to be difficult to implement in Go, I would say that the cost is not worth the benefits.  But I don't believe a global map would be problematic at all to implement, so I would be fine with tracking it globally.  That said, I also don't think it's likely to make much of a difference either way, since _most of the time_, "the current process" will devolve into a single client anyway.

Reply:
Thanks, that was really useful discussion!

@ejona86, to me, the key point in what you said was actually this:
> If we take a least-request policy with a fixed set of addresses and duplicate the entire thing N times (duplicating the counters as well), then the behavior of those N policies should actually be fairly similar to the single unified policy.

I think you're right about the various edge cases with low number of concurrent requests or different mixes of differently weighted types of requests, but those things can happen even across client processes or machines or any other scope we could consider.  So I don't think we need to really worry about those, and I agree with your point that in most situations where this policy will be used, the behavior should be the same no matter what scope we choose here.

So, I think we should stick with a purely local-to-the-LB-policy view.  Even if it turns out that C-core will want to put this in a map keyed by subchannel ID rather than storing it in the subchannel wrappers, the fact that that map isn't global means that we don't need a bunch of additional synchronization, which seems like a win.

@erikjoh, I think we can change this gRFC to say that the counters can be stored in a wrapped subchannel.  In C-core, there will likely be some indirection there -- the counters themselves will probably be in a separate map, and the wrapped subchannel will hold a ref to the relevant entry in that map -- but either way, the wrapped subchannel is the means that the picker will use to determine which counter to increment for a given request, so I think that's the important thing to note here.

Reply:
Very informative discussion indeed!
@markdroth I've added a few sentences about this in the outstanding counters section and also moved that section before the picker section since it seems like a more logical ordering here.
I was a bit careful with the wording since I don't think the Java implementation actually needs to wrap the subchannels since subchannels are already LB-local and also the support for custom subchannel attributes can be utilized to store the counters.

Reply:
That's a good point.  I was thinking about the fact that you will need to create an object that contains both the subchannel and the counter for a given address, and the picker will need to access that object so that it can increment the counter when it returns the subchannel.  But you're right that that object doesn't necessarily need to wrap the subchannel, since this is a leaf policy and won't be passing the subchannel down to a child policy and then unwrapping it when it comes back from the child's picker.